optimade/validator/validator.py

```diff
@@ -390,34 +445,58 @@ def main(self):
         self._log.debug("Testing %s endpoint", LINKS_ENDPOINT)
         self.test_info_or_links_endpoints(LINKS_ENDPOINT)
 
-        self.valid = not bool(self.failure_count)
+        self.valid = not (bool(self.failure_count) or bool(self.internal_failure_count))
 
-        print("\nOptional tests:")
-        for endp in self.endpoint_optional_queries:
-            # skip empty endpoint query lists
-            if self.endpoint_mandatory_queries[endp]:
-                self._log.debug("Testing optional query syntax on endpoint %s", endp)
-                self.test_query_syntax(
-                    endp, self.endpoint_optional_queries[endp], optional=True
-                )
+        if self.run_optional_tests:
+            print("\nOptional tests:")
+            for endp in self.endpoint_optional_queries:
+                # skip empty endpoint query lists
+                if self.endpoint_mandatory_queries[endp]:
+                    self._log.debug(
+                        "Testing optional query syntax on endpoint %s", endp
+                    )
+                    self.test_query_syntax(
+                        endp, self.endpoint_optional_queries[endp], optional=True
+                    )
 
-        if not self.valid:
-            print("\n\nFAILURES\n")
-            for message in self.failure_messages:
-                print_failure(message[0])
-                for line in message[1]:
-                    print_warning("\t" + line)
+        self.print_summary()
 
-        final_message = f"\n\nPassed {self.success_count} out of {self.success_count + self.failure_count} tests."
+    def print_summary(self):
         if not self.valid:
-            print_failure(final_message)
-        else:
-            print_success(final_message)
+            if self.failure_messages:
+                print("\n\nFAILURES")
+                print("========\n")
+                for message in self.failure_messages:
+                    print_failure(message[0])
+                    for line in message[1]:
+                        print_warning("\t" + line)
+
+            if self.internal_failure_messages:
+                print("\n\nINTERNAL FAILURES")
+                print("=================\n")
+                print(
+                    "There were internal valiator failures associated with this run.\n"
+                    "If this problem persists, please report it at:\n"
+                    "https://github.com/Materials-Consortia/optimade-python-tools/issues/new.\n"
+                )
 
-        print(
-            f"Additionally passed {self.optional_success_count} out of "
-            f"{self.optional_success_count + self.optional_failure_count} optional tests."
-        )
+                for message in self.internal_failure_messages:
+                    print_notify(message[0])
+                    for line in message[1]:
+                        print_warning("\t" + line)
+
+        if not self.valid and not self.fail_fast:
```

Comment:
```suggestion
        if not self.fail_fast:
```

Reply:
Admit this is a bit clumsy, but think the suggestion changes the logic. Will go for something like `if self.valid or (not self.valid and self.fail_fast):` which is verbose but I think clearer?

Reply:
Well if it need to have that addition, why not put it in the previous `if`-block? I thought it was a mistype, since within this `if`-block you're retesting `self.valid`, which doesn't make sense since it will never reach it unless `not self.valid is True`.

Reply:
> Admit this is a bit clumsy, but think the suggestion changes the logic. Will go for something like `if self.valid or (not self.valid and not self.fail_fast):` which is verbose but I think clearer?

Also, how does your suggestion not also change the logic of the original condition? :)

Reply:
The block should only execute if `(valid, fail_fast)` is `(False, False)`, which I think was fine initially? The suggestion just spells out the fact that it will be executed in the case of `(True, False)`. `fail_fast` is the user-provided option of whether to `fail_fast` in the case of an invalid response (rather than indicating that things have already failed) which I think makes it seem a bit confusing...

Reply:
Huh. Well I'm still confused 😅 
So the `if`-block should execute if `(valid, fail_fast) == (False, False)`, but you want add that it runs always also if `valid is True`?
Or, if indeed it still should execute only if `(valid, fail_fast) == (False, False)`, then why not indent it into the previous `if`-block, that runs when `valid is False`?

Reply:
Just realised how confusing my last comment was, I meant that it should also execute on `valid=True`, no matter the value of `fail_fast`, so it's fine as it is now...